### PR TITLE
Add new query state "spooling"

### PR DESCRIPTION
**Pull request by chi-tsai**
Test plan - Unit tests

```
== RELEASE NOTES ==

General Changes
* Add new state 'spooling' for queries that has finished fetching the result, and are only waiting for the client to retrieve the data.
```

### CODE REVIEW

**Review by tdcmeehan — COMMENTED** [review]
I think we should reconsider the way we calculate `spoolingTime` in `QueryStateTimer`, and just calculate it like we do earlier state transitions.  The code changes to accomodate the logic where this is calculated at the end are looking to be kind of convoluted.

**Comment by tdcmeehan** [comment]
Re: test plan, it will be good to get verifier and shadow tests on this too; we should cherry-pick several queries to make sure we're not messing up the query and task stats.

File: presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java [hunk 1]
```diff
@@ -644,7 +645,10 @@ private void addStateChangeListeners(SectionExecution sectionExecution)
             SqlStageExecution stageExecution = stageExecutionAndScheduler.getStageExecution();
             if (isRootFragment(stageExecution.getFragment())) {
                 stageExecution.addStateChangeListener(state -> {
-                    if (state == FINISHED) {
+                    if (state == SPOOLING) {
+                        queryStateMachine.transitionToSpooling();
```
Comment (swapsmagic) [inline presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java:649]: ditto

**Review by swapsmagic — COMMENTED** [review]
Add tests to make sure queries are not being counted from the resource group if moved to spooling state. You can add those tests in TestQueuesDb class.

File: presto-main/src/main/java/com/facebook/presto/execution/SqlStageExecution.java [hunk 2]
```diff
@@ -289,6 +289,11 @@ public synchronized void schedulingComplete()
         if (getAllTasks().stream().anyMatch(task -> getState() == StageExecutionState.RUNNING)) {
             stateMachine.transitionToRunning();
         }
+
+        if (getAllTasks().stream().allMatch(task -> task.getTaskStatus().getState() == TaskState.SPOOLING || task.getTaskStatus().getState() == TaskState.FINISHED)) {
```
[the remark this answers is not in the document]
Comment (tdcmeehan) [reply presto-main/src/main/java/com/facebook/presto/execution/SqlStageExecution.java:293]: +1

File: presto-main/src/main/java/com/facebook/presto/execution/SqlTaskExecution.java [hunk 3]
```diff
@@ -643,6 +644,7 @@ private synchronized void checkTaskCompletion()
 
         // are there still pages in the output buffer?
         if (!outputBuffer.isFinished()) {
+            taskStateMachine.spooling();
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/execution/SqlTaskExecution.java:647]: Let's guard this with a session property.

**Review by tdcmeehan — APPROVED** [review]
LGTM, let's just be sure to guard the task transition by a session property, and make sure stage/section/query don't transition to spooling if the dependent stages/sections/tasks don't themselves transition to spooling.

File: presto-main/src/main/java/com/facebook/presto/execution/scheduler/LegacySqlQueryScheduler.java [hunk 4]
```diff
@@ -253,7 +254,10 @@ private void initialize()
     {
         SqlStageExecution rootStage = stageExecutions.get(rootStageId).getStageExecution();
         rootStage.addStateChangeListener(state -> {
-            if (state == FINISHED) {
+            if (state == SPOOLING) {
```
[the remark this answers is not in the document]
Comment (swapsmagic) [reply presto-main/src/main/java/com/facebook/presto/execution/scheduler/LegacySqlQueryScheduler.java:257]: nvmd, found out that the root state won't move to SPOOLING till all other states finished processing. So it's good.

File: presto-main/src/main/java/com/facebook/presto/execution/scheduler/LegacySqlQueryScheduler.java [hunk 5]
```diff
@@ -674,7 +678,10 @@ private void addStateChangeListeners(SectionExecution sectionExecution)
             SqlStageExecution stageExecution = stageExecutionAndScheduler.getStageExecution();
             if (isRootFragment(stageExecution.getFragment())) {
                 stageExecution.addStateChangeListener(state -> {
-                    if (state == FINISHED) {
+                    if (state == SPOOLING) {
+                        queryStateMachine.transitionToSpooling();
```
[the remark this answers is not in the document]
Comment (swapsmagic) [reply presto-main/src/main/java/com/facebook/presto/execution/scheduler/LegacySqlQueryScheduler.java:682]: ignore

File: presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java (hunk 1 above)
Comment (swapsmagic) [reply presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java:649]: ignore

**Pull request closed without merging**